dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/TimeToLive.java

```diff
@@ -0,0 +1,154 @@
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.SIX_MONTHS;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.WEEKLY;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_MONTHLY_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_QUARTERLY_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_SIX_MONTHS_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_WEEKLY_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_YEARLY_OR_OVER_PERIOD_IN_SECONDS;
+import static org.springframework.util.Assert.notNull;
+
+import java.time.Instant;
+
+import org.hisp.dhis.analytics.DataQueryParams;
+import org.hisp.dhis.setting.SettingKey;
+import org.hisp.dhis.setting.SystemSettingManager;
+
+public class TimeToLive
+    implements
+    Computable
+{
+
+    static final long DEFAULT_TO_30_SECONDS = 30;
+
+    private final DataQueryParams params;
+
+    private SystemSettingManager systemSettingManager;
+
+    public TimeToLive( final DataQueryParams params, final SystemSettingManager systemSettingManager )
+    {
+        notNull( params, "Object params must not be null" );
+        notNull( params.getEarliestStartDate(), "Object params.getEarliestStartDate() must not be null" );
+        notNull( params.getLatestEndDate(), "Object params.getLatestEndDate() must not be null" );
+        notNull( systemSettingManager, "Object systemSettingManager must not be null" );
+        this.params = params;
+        this.systemSettingManager = systemSettingManager;
+    }
+
+    /**
+     * Calculates the time to live based on specific internal rules.
+     *
+     * @return the computed TTL value in SECONDS.
+     */
+    @Override
+    public long compute()
+    {
+        final Instant oldestDate = params.getEarliestStartDate().toInstant();
+        final Instant mostRecentDate = params.getLatestEndDate().toInstant();
+        final long daysDiff = DAYS.between( oldestDate, mostRecentDate );
+        return calculationFor( daysDiff, mostRecentDate );
+    }
+
+    /**
+     * Execute the internal rules in order to calculate a TTL for the given
+     * parameters.
+     *
+     * @param daysDiff the difference of days between the oldest and most recent
+     *        date in the period.
+     * @param mostRecentDate the most recent date of the period.
+     * @return a time to live value in SECONDS.
+     */
+    long calculationFor( final long daysDiff, final Instant mostRecentDate )
+    {
+        /*
+         * If the difference between the most recent date and NOW is 0 (zero) it means
+         * the current year, so we increment the multiplier by 1 (one) avoiding
+         * multiplying by 0 (zero).
+         */
+        final long ttlMultiplierOrOne = YEARS.between( ofInstant( mostRecentDate, UTC ), now() ) + 1;
```

Comment:
Is "orOne" not entirely correct as we always add one to the calculated value (not just when zero) ?

Reply:
done